rasterio/rio/features.py

```diff
@@ -20,6 +19,128 @@
 logger = logging.getLogger('rio')
 
 
+# Extract command
+@cli.command(short_help='Extract raster using features.')
+@files_inout_arg
+@format_opt
+@click.option('--all_touched', is_flag=True, default=False,
+              help='Use all pixels touched by features for extraction, '
+              'otherwise use only pixels whose center is within the '
+              'polygon or that are selected by Brezenhams line algorithm')
+@click.option('--crop', is_flag=True, default=False,
+              help='Crop output raster to the extent of the geometries. '
+                   'GeoJSON must overlap input raster to use --crop')
+@click.option('--invert', is_flag=True, default=False,
+              help='Inverts the mask, so that areas covered by features are'
+                   'masked out and areas not covered are retained.  Ignored '
+                   'if using --crop')
+@click.pass_context
+def extract(
+        ctx,
+        files,
+        driver,
+        all_touched,
+        crop,
+        invert):
+
+    """Extracts pixels from all bands of a raster using features, masking out
+    all areas not covered by features, and optionally crops the output raster
+    to the extent of the features.  Features are assumed to be in the same
+    coordinate reference system as the input raster.
+
+    GeoJSON must be the first input file or provided from stdin:
+
+    > rio extract features.json input.tif output.tif
+
+    > rio extract input.tif output.tif < features.json
+
+    If the output raster exists, it will be completely overwritten with the
+    results of this operation.
+
+    The result is always equal to or within the bounds of the input raster.
+
+    --crop and --invert options are mutually exclusive.
+
+    --crop option is not valid if features are completely outside extent of
+    input raster.
+    """
```

Comment:
@brendan-ward This program strikes me as being a filter like `grep`. What would you think about borrowing command line options from grep? `--invert` could become `-v, --invert-mask` (like grep's `-v, --invert-match`). And I'd like to see the GeoJSON be more like grep's `pattern`: not an input filename but either a GeoJSON text argument or an `-f/--file features.json` option. If there were no features, we could pass all the input pixels through.

So, the canonical example might be 

```
$ rio extract -f features.json input.tif output.tif
```



```
     
             Selected lines are those not matching any of the specified pat-
             terns.
```